File: 2022/day04/day04.jl
module Day04

function contained(a, b)
    return (a[1] <= b[1] && a[2] >= b[2]) || (a[1] >= b[1] && a[2] <= b[2])
end

function overlaps(a, b)
    (a[2] >= b[1] && b[2] >= a[1]) || (b[2] >= a[1] && a[2] >= b[1])
end

function solver(check, input)
    total = 0
    for line in eachline(input)
        pair1, pair2 = split.(split(line, ","), "-")
        if check((parse(Int, pair1[1]), parse(Int, pair1[2])), (parse(Int, pair2[1]), parse(Int, pair2[2])))
            total += 1
        end
    end
    println(total)
end

function part1(input = "input.txt")
    solver(contained, input)
end

function part2(input = "input.txt")
    solver(overlaps, input)
end

end

File: 2023/elixir/src/day04.ex
defmodule Day04 do
  def get_intersecting(line) do
    [first, last] =
      String.replace(line, ~r/Card\s+\d+: /, "")
      |> String.split("|")
      |> Enum.map(fn group ->
        String.split(group)
        |> Enum.map(fn match -> String.to_integer(match) end)
        |> MapSet.new()
      end)

    MapSet.intersection(first, last) |> MapSet.size()
  end

  def part1(input) do
    input
    |> String.split("\n")
    |> Enum.map(fn line ->
      n = get_intersecting(line)
      if n == 0, do: 0, else: 2 ** (n - 1)
    end)
    |> Enum.sum()
  end

  def part2(input) do
    lines = String.split(input, "\n")

    lines
    |> Enum.with_index()
    |> Enum.reduce(List.duplicate(1, length(lines)), fn {line, i}, acc ->
      n = get_intersecting(line)

      n_cards = elem(List.to_tuple(acc), i)

      if n == 0 do
        acc
      else
        acc
        |> Enum.with_index()
        |> Enum.map(fn {el, j} ->
          if i < j && j <= i + n do
            el + n_cards
          else
            el
          end
        end)
      end
    end)
    |> Enum.sum()
  end
end

input = File.read!("inputs/day04.txt")

Day04.part1(input) |> IO.inspect()
Day04.part2(input) |> IO.inspect()
